File: examples/python/servers/mcp/advanced.py
"""MCP Server with x402 Paid Tools - Advanced Example with Hooks.

This example demonstrates using create_payment_wrapper with hooks for:
- Logging and observability
- Rate limiting and access control
- Custom settlement handling
- Production monitoring

Run with: python main.py advanced
"""

import json
import os
import random
import sys
from typing import Any

from dotenv import load_dotenv
from mcp.server.fastmcp import FastMCP

from x402.http import FacilitatorConfig, HTTPFacilitatorClient
from x402.mechanisms.evm.exact import ExactEvmServerScheme
from x402.mcp import (
    AfterExecutionContext,
    MCPContentItem,
    MCPToolContext,
    MCPToolResult,
    PaymentWrapperConfig,
    PaymentWrapperHooks,
    ResourceInfo,
    ServerHookContext,
    SettlementContext,
    create_payment_wrapper,
)
from x402.schemas import ResourceConfig
from x402.server import x402ResourceServerSync

load_dotenv()

EVM_ADDRESS = os.getenv("EVM_ADDRESS")
FACILITATOR_URL = os.getenv("FACILITATOR_URL", "https://x402.org/facilitator")
PORT = int(os.getenv("PORT", "4022"))

if not EVM_ADDRESS:
    print("❌ EVM_ADDRESS environment variable is required")
    sys.exit(1)



def get_weather_data(city: str) -> dict[str, Any]:
    """Simulate fetching weather data for a city."""
    conditions = ["sunny", "cloudy", "rainy", "snowy", "windy"]
    weather = random.choice(conditions)
    temperature = random.randint(40, 80)
    return {"city": city, "weather": weather, "temperature": temperature}


def run_advanced() -> None:
    """Main entry point - demonstrates hooks with payment wrapper using REAL MCP SDK."""
    print("\n📦 Using Payment Wrapper with Hooks and REAL MCP SDK\n")

    # ========================================================================
    # STEP 1: Create REAL MCP server using FastMCP
    # ========================================================================
    mcp_server = FastMCP("x402-mcp-server-advanced", json_response=True)

    facilitator_client = HTTPFacilitatorClient(FacilitatorConfig(url=FACILITATOR_URL))
    resource_server = x402ResourceServerSync(facilitator_client)
    resource_server.register("eip155:84532", ExactEvmServerScheme())
    resource_server.initialize()

    # Build payment requirements
    weather_config = ResourceConfig(
        scheme="exact",
        network="eip155:84532",
        pay_to=EVM_ADDRESS,
        price="$0.001",
        extra={"name": "USDC", "version": "2"},
    )
    weather_accepts = resource_server.build_payment_requirements(weather_config)

    forecast_config = ResourceConfig(
        scheme="exact",
        network="eip155:84532",
        pay_to=EVM_ADDRESS,
        price="$0.005",
        extra={"name": "USDC", "version": "2"},
    )
    forecast_accepts = resource_server.build_payment_requirements(forecast_config)

    # Shared hooks
    def before_hook(context: ServerHookContext) -> tuple[bool, None]:
        print(f"\n🔧 [Hook] Before execution: {context.tool_name}")
        if context.payment_payload.payload:
            auth = context.payment_payload.payload.get("authorization", {})
            if isinstance(auth, dict):
                from_addr = auth.get("from")
                if from_addr:
                    print(f"   Payment from: {from_addr}")
        print(f"   Amount: {context.payment_requirements.amount}")
        return True, None

    def after_hook(context: AfterExecutionContext) -> None:
        print(f"✅ [Hook] After execution: {context.tool_name}")
        print(f"   Result error: {context.result.is_error}")

    def settlement_hook(context: SettlementContext) -> None:
        print(f"💸 [Hook] Settlement complete: {context.tool_name}")
        if context.settlement.transaction:
            print(f"   Transaction: {context.settlement.transaction}")
        print(f"   Success: {context.settlement.success}\n")

    shared_hooks = PaymentWrapperHooks(
        on_before_execution=before_hook,
        on_after_execution=after_hook,
        on_after_settlement=settlement_hook,
    )

    paid_weather = create_payment_wrapper(
        resource_server,
        PaymentWrapperConfig(
            accepts=weather_accepts,
            resource=ResourceInfo(
                url="mcp://tool/get_weather",
                description="Get weather for a city",
                mime_type="application/json",
            ),
            hooks=shared_hooks,
        ),
    )

    paid_forecast = create_payment_wrapper(
        resource_server,
        PaymentWrapperConfig(
            accepts=forecast_accepts,
            resource=ResourceInfo(
                url="mcp://tool/get_forecast",
                description="Get 7-day forecast",
                mime_type="application/json",
            ),
            hooks=shared_hooks,
        ),
    )

    # ========================================================================
    # STEP 5: Register tools using REAL MCP SDK with hooks
    # ========================================================================

    # Free tool
    @mcp_server.tool()
    def ping() -> str:
        """A free health check tool."""
        return "pong"

    # Weather tool with hooks
    @mcp_server.tool()
    def get_weather(city: str = "San Francisco") -> str:
        """Get current weather for a city. Requires payment of $0.001."""
        # Call paid handler with payment wrapper
        result = paid_weather(
            lambda args, tc: MCPToolResult(
                content=[MCPContentItem(type="text", text=json.dumps(get_weather_data(city), indent=2))],
                is_error=False,
            )
        )({"city": city}, {"_meta": {}, "toolName": "get_weather"})

        if result.is_error:
            if result.content and len(result.content) > 0:
                return result.content[0].text if isinstance(result.content[0], MCPContentItem) else result.content[0].get("text", "Payment required")
            return "Payment required"

        return result.content[0].text if result.content and isinstance(result.content[0], MCPContentItem) else (result.content[0].get("text", "Weather data") if result.content else "No data")

    # Forecast tool with hooks
    @mcp_server.tool()
    def get_forecast(city: str = "San Francisco") -> str:
        """Get 7-day weather forecast. Requires payment of $0.005."""
        forecast = [
            {**get_weather_data(city), "day": i + 1} for i in range(7)
        ]

        # Call paid handler with payment wrapper
        result = paid_forecast(
            lambda args, tc: MCPToolResult(
                content=[MCPContentItem(type="text", text=json.dumps(forecast, indent=2))],
                is_error=False,
            )
        )({"city": city}, {"_meta": {}, "toolName": "get_forecast"})

        if result.is_error:
            if result.content and len(result.content) > 0:
                return result.content[0].text if isinstance(result.content[0], MCPContentItem) else result.content[0].get("text", "Payment required")
            return "Payment required"

        return result.content[0].text if result.content and isinstance(result.content[0], MCPContentItem) else (result.content[0].get("text", "Forecast data") if result.content else "No data")

    print(f"🚀 x402 MCP Server (Advanced) running on http://localhost:{PORT}")
    print("\n📋 Available tools:")
    print("   - get_weather (paid: $0.001) [with hooks]")
    print("   - get_forecast (paid: $0.005) [with hooks]")
    print("   - ping (free)")
    print(f"\n🔗 Connect via streamable HTTP: http://localhost:{PORT}/mcp")
    print("\n📊 Hooks enabled:")
    print("   - OnBeforeExecution: Rate limiting, validation")
    print("   - OnAfterExecution: Logging, metrics")
    print("   - OnAfterSettlement: Receipts, notifications")
    print("\n💡 This example uses create_payment_wrapper() with hooks and REAL MCP SDK.\n")

    # Run server with streamable HTTP transport
    mcp_server.run(transport="streamable-http", port=PORT)

File: examples/go/servers/mcp/advanced.go
package main

// MCP Server with x402 Paid Tools - Advanced Example with Hooks
//
// This example demonstrates using CreatePaymentWrapper with hooks for:
// - Logging and observability
// - Rate limiting and access control
// - Custom settlement handling
// - Production monitoring
//
// The getWeatherData helper is defined in helpers.go and shared across examples.
//
// Run with: go run . advanced

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	x402 "github.com/coinbase/x402/go"
	x402http "github.com/coinbase/x402/go/http"
	evm "github.com/coinbase/x402/go/mechanisms/evm/exact/server"
	"github.com/coinbase/x402/go/mcp"
	"github.com/joho/godotenv"
	mcpsdk "github.com/modelcontextprotocol/go-sdk/mcp"
)

func runAdvanced() error {
	fmt.Println("\n📦 Using Payment Wrapper with Hooks\n")

	// Load environment variables
	if err := godotenv.Load(); err != nil {
		fmt.Println("No .env file found, using environment variables")
	}

	evmAddress := os.Getenv("EVM_ADDRESS")
	if evmAddress == "" {
		return fmt.Errorf("EVM_ADDRESS environment variable is required")
	}

	facilitatorURL := os.Getenv("FACILITATOR_URL")
	if facilitatorURL == "" {
		return fmt.Errorf("FACILITATOR_URL environment variable is required")
	}

	port := os.Getenv("PORT")
	if port == "" {
		port = "4022"
	}

	// ========================================================================
	// STEP 1: Create REAL MCP server
	// ========================================================================
	mcpServer := mcpsdk.NewServer(&mcpsdk.Implementation{
		Name:    "x402 MCP Server (Advanced)",
		Version: "1.0.0",
	}, nil)

	// ========================================================================
	// STEP 2: Set up x402 resource server
	// ========================================================================
	ctx := context.Background()
	facilitatorClient := x402http.NewHTTPFacilitatorClient(&x402http.FacilitatorConfig{
		URL: facilitatorURL,
	})

	resourceServer := x402.Newx402ResourceServer(
		x402.WithFacilitatorClient(facilitatorClient),
	)
	resourceServer.Register("eip155:84532", evm.NewExactEvmScheme())

	if err := resourceServer.Initialize(ctx); err != nil {
		return fmt.Errorf("failed to initialize resource server: %w", err)
	}

	// ========================================================================
	// STEP 3: Build payment requirements for different tools
	// ========================================================================
	weatherConfig := x402.ResourceConfig{
		Scheme:  "exact",
		Network: "eip155:84532",
		PayTo:   evmAddress,
		Price:   "$0.001",
	}

	weatherAccepts, err := resourceServer.BuildPaymentRequirementsFromConfig(ctx, weatherConfig)
	if err != nil {
		return fmt.Errorf("failed to build weather payment requirements: %w", err)
	}

	forecastConfig := x402.ResourceConfig{
		Scheme:  "exact",
		Network: "eip155:84532",
		PayTo:   evmAddress,
		Price:   "$0.005",
	}

	forecastAccepts, err := resourceServer.BuildPaymentRequirementsFromConfig(ctx, forecastConfig)
	if err != nil {
		return fmt.Errorf("failed to build forecast payment requirements: %w", err)
	}

	// ========================================================================
	// STEP 4: Create payment wrappers with hooks for production features
	// ========================================================================

	// Shared hooks for all paid tools
	beforeHook := mcp.BeforeExecutionHook(func(context mcp.ServerHookContext) (bool, error) {
		fmt.Printf("\n🔧 [Hook] Before execution: %s\n", context.ToolName)
		if context.PaymentPayload.Payload != nil {
			if auth, ok := context.PaymentPayload.Payload["authorization"].(map[string]interface{}); ok {
				if from, ok := auth["from"].(string); ok {
					fmt.Printf("   Payment from: %s\n", from)
				}
			}
		}
		fmt.Printf("   Amount: %s\n", context.PaymentRequirements.Amount)
		return true, nil // Continue execution
	})

	afterHook := mcp.AfterExecutionHook(func(context mcp.AfterExecutionContext) error {
		fmt.Printf("✅ [Hook] After execution: %s\n", context.ToolName)
		fmt.Printf("   Result error: %v\n", context.Result.IsError)
		return nil
	})

	settlementHook := mcp.AfterSettlementHook(func(context mcp.SettlementContext) error {
		fmt.Printf("💸 [Hook] Settlement complete: %s\n", context.ToolName)
		if context.Settlement.Transaction != "" {
			fmt.Printf("   Transaction: %s\n", context.Settlement.Transaction)
		}
		fmt.Printf("   Success: %v\n\n", context.Settlement.Success)
		return nil
	})

	sharedHooks := &mcp.PaymentWrapperHooks{
		OnBeforeExecution: &beforeHook,
		OnAfterExecution:  &afterHook,
		OnAfterSettlement: &settlementHook,
	}

	paidWeather, err := mcp.CreatePaymentWrapper(resourceServer, mcp.PaymentWrapperConfig{
		Accepts: weatherAccepts,
		Resource: &mcp.ResourceInfo{
			URL:         "mcp://tool/get_weather",
			Description: "Get weather for a city",
			MimeType:    "application/json",
		},
		Hooks: sharedHooks,
	})
	if err != nil {
		return fmt.Errorf("failed to create weather payment wrapper: %w", err)
	}

	paidForecast, err := mcp.CreatePaymentWrapper(resourceServer, mcp.PaymentWrapperConfig{
		Accepts: forecastAccepts,
		Resource: &mcp.ResourceInfo{
			URL:         "mcp://tool/get_forecast",
			Description: "Get 7-day forecast",
			MimeType:    "application/json",
		},
		Hooks: sharedHooks,
	})
	if err != nil {
		return fmt.Errorf("failed to create forecast payment wrapper: %w", err)
	}

	// ========================================================================
	// STEP 5: Register tools using REAL MCP SDK
	// ========================================================================

	// Free tool
	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "ping",
		Description: "A free health check tool",
		InputSchema: map[string]interface{}{"type": "object"},
	}, func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		return &mcpsdk.CallToolResult{
			Content: []mcpsdk.Content{
				&mcpsdk.TextContent{Text: "pong"},
			},
		}, nil
	})

	// Weather tool - $0.001 with hooks
	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "get_weather",
		Description: "Get current weather for a city. Requires payment of $0.001.",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"city": map[string]interface{}{"type": "string", "description": "The city name"},
			},
		},
	}, func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		args := make(map[string]interface{})
		if req.Params.Arguments != nil {
			if err := json.Unmarshal(req.Params.Arguments, &args); err != nil {
				args = make(map[string]interface{})
			}
		}
		meta := make(map[string]interface{})
		if req.Params.Meta != nil {
			meta = req.Params.Meta.GetMeta()
		}

		toolContext := mcp.MCPToolContext{
			ToolName:  req.Params.Name,
			Arguments: args,
			Meta:      meta,
		}

		city, _ := args["city"].(string)
		if city == "" {
			city = "San Francisco"
		}

		result, err := paidWeather(func(ctx context.Context, args map[string]interface{}, toolContext mcp.MCPToolContext) (mcp.MCPToolResult, error) {
			weatherData := getWeatherData(city)
			weatherJSON, _ := json.MarshalIndent(weatherData, "", "  ")
			return mcp.MCPToolResult{
				Content: []mcp.MCPContentItem{
					{Type: "text", Text: string(weatherJSON)},
				},
				IsError: false,
			}, nil
		})(ctx, args, toolContext)

		if err != nil {
			return &mcpsdk.CallToolResult{
				IsError: true,
				Content: []mcpsdk.Content{
					&mcpsdk.TextContent{Text: err.Error()},
				},
			}, nil
		}

		content := make([]mcpsdk.Content, len(result.Content))
		for i, item := range result.Content {
			content[i] = &mcpsdk.TextContent{Text: item.Text}
		}

		callResult := &mcpsdk.CallToolResult{
			Content: content,
			IsError: result.IsError,
		}

		if result.Meta != nil {
			callResult.Meta = mcpsdk.Meta(result.Meta)
		}

		return callResult, nil
	})

	// Forecast tool - $0.005 with hooks
	mcpServer.AddTool(&mcpsdk.Tool{
		Name:        "get_forecast",
		Description: "Get 7-day weather forecast. Requires payment of $0.005.",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"city": map[string]interface{}{"type": "string", "description": "The city name"},
			},
		},
	}, func(ctx context.Context, req *mcpsdk.CallToolRequest) (*mcpsdk.CallToolResult, error) {
		args := make(map[string]interface{})
		if req.Params.Arguments != nil {
			if err := json.Unmarshal(req.Params.Arguments, &args); err != nil {
				args = make(map[string]interface{})
			}
		}
		meta := make(map[string]interface{})
		if req.Params.Meta != nil {
			meta = req.Params.Meta.GetMeta()
		}

		toolContext := mcp.MCPToolContext{
			ToolName:  req.Params.Name,
			Arguments: args,
			Meta:      meta,
		}

		city, _ := args["city"].(string)
		if city == "" {
			city = "San Francisco"
		}

		result, err := paidForecast(func(ctx context.Context, args map[string]interface{}, toolContext mcp.MCPToolContext) (mcp.MCPToolResult, error) {
			forecast := make([]map[string]interface{}, 7)
			for i := 0; i < 7; i++ {
				dayData := getWeatherData(city)
				dayData["day"] = i + 1
				forecast[i] = dayData
			}
			forecastJSON, _ := json.MarshalIndent(forecast, "", "  ")
			return mcp.MCPToolResult{
				Content: []mcp.MCPContentItem{
					{Type: "text", Text: string(forecastJSON)},
				},
				IsError: false,
			}, nil
		})(ctx, args, toolContext)

		if err != nil {
			return &mcpsdk.CallToolResult{
				IsError: true,
				Content: []mcpsdk.Content{
					&mcpsdk.TextContent{Text: err.Error()},
				},
			}, nil
		}

		content := make([]mcpsdk.Content, len(result.Content))
		for i, item := range result.Content {
			content[i] = &mcpsdk.TextContent{Text: item.Text}
		}

		callResult := &mcpsdk.CallToolResult{
			Content: content,
			IsError: result.IsError,
		}

		if result.Meta != nil {
			callResult.Meta = mcpsdk.Meta(result.Meta)
		}

		return callResult, nil
	})

	// Start HTTP server with SSE transport
	return startHTTPServerAdvanced(mcpServer, port)
}

func startHTTPServerAdvanced(mcpServer *mcpsdk.Server, port string) error {
	sseHandler := mcpsdk.NewSSEHandler(func(req *http.Request) *mcpsdk.Server {
		return mcpServer
	}, nil)

	mux := http.NewServeMux()
	mux.Handle("/sse", sseHandler)
	mux.Handle("/messages", sseHandler)

	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"status": "ok",
			"mode":   "advanced-with-hooks",
			"server": "x402 MCP Server (Advanced)",
		})
	})

	fmt.Printf("🚀 x402 MCP Server (Advanced) running on http://localhost:%s\n", port)
	fmt.Println("\n📋 Available tools:")
	fmt.Println("   - get_weather (paid: $0.001) [with hooks]")
	fmt.Println("   - get_forecast (paid: $0.005) [with hooks]")
	fmt.Println("   - ping (free)")
	fmt.Printf("\n🔗 Connect via SSE: http://localhost:%s/sse\n", port)
	fmt.Println("\n📊 Hooks enabled:")
	fmt.Println("   - OnBeforeExecution: Rate limiting, validation")
	fmt.Println("   - OnAfterExecution: Logging, metrics")
	fmt.Println("   - OnAfterSettlement: Receipts, notifications")
	fmt.Println("\n💡 This example uses CreatePaymentWrapper() with hooks and REAL MCP SDK.\n")

	server := &http.Server{
		Addr:    ":" + port,
		Handler: mux,
	}

	return server.ListenAndServe()
}
